File: 104_maximum_depth_of_binary_tree/maxDepth.py
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        
        def dfs(root: Optional[TreeNode]) -> int:
            if root == None:
                return 0

            left = dfs(root.left)
            right = dfs(root.right)

            return max(left, right) + 1

        return dfs(root)

File: 104_maximum_depth_of_binary_tree/maxDepth.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int depth;

    private int dfs(TreeNode root) {
        if (root == null) return 0;

        int left = dfs(root.left);
        int right = dfs(root.right);
        int d = Math.max(left, right) + 1;
        
        depth = Math.max(depth, d);
        return d;
    }
    public int maxDepth(TreeNode root) {
        dfs(root);

        return depth;
    }
}